adr-11-centralized-sequencer.md

```diff
@@ -0,0 +1,223 @@
+
+### Embedded Sequencer in Rollup Nodes
+
+Another approach would be to embed sequencing functionality directly into rollup nodes. This would simplify the architecture by eliminating a separate sequencer component.
+
+This approach was not chosen because:
+1. It couples sequencing logic with rollup node logic, reducing modularity
+2. It makes it harder to upgrade or replace the sequencing component independently
+3. It doesn't allow for a dedicated sequencing service that can be optimized separately
+
+## Decision
+
+We implement a standalone centralized sequencer that:
+
+1. Implements the Generic Sequencer interface from the `go-sequencing` package
+2. Exposes a gRPC service for rollup clients to submit transactions
+3. Batches transactions and submits them to a DA layer at regular intervals
+4. Maintains state to track submitted batches and provide verification
+5. Provides metrics for monitoring and observability
+
+The centralized sequencer is a separate repository and can be deployed as a standalone service or as a Docker container.
+
+## Detailed Design
+
+### User Requirements
+
+- Rollup developers need a simple, reliable sequencer that can order transactions and submit them to a DA layer
+- The sequencer should be easy to deploy and configure
+- The sequencer should provide metrics for monitoring
+- The sequencer should be able to recover from crashes and maintain state
+
+### Systems Affected
+
+- Rollup nodes that interact with the sequencer
+- DA layer where batches are submitted
+
+### Data Structures
+
+The centralized sequencer uses the following key data structures:
+
+1. **BatchQueue**: A queue to store batches of transactions waiting to be processed
+   ```go
+   type BatchQueue struct {
+       queue []sequencing.Batch  // In-memory queue of batches waiting to be processed
+       mu    sync.Mutex          // Mutex to ensure thread-safe access to the queue
+   }
+   ```
+
+2. **Sequencer**: The main sequencer structure that implements the Generic Sequencer interface
+   ```go
+   type Sequencer struct {
+       dalc      *da.DAClient       // Client for interacting with the Data Availability layer
+       batchTime time.Duration      // Time interval between batch submissions
+       ctx       context.Context    // Context for controlling the sequencer's lifecycle
+       maxSize   uint64             // Maximum size of a batch in bytes
+
+       rollupId sequencing.RollupId // Identifier for the rollup this sequencer serves
+
+       tq                 *TransactionQueue  // Queue for storing pending transactions
+       lastBatchHash      []byte             // Hash of the last processed batch
+       lastBatchHashMutex sync.RWMutex       // Mutex for thread-safe access to lastBatchHash
+
+       seenBatches      map[string]struct{}  // Map to track batches that have been processed
+       seenBatchesMutex sync.Mutex           // Mutex for thread-safe access to seenBatches
```

Comment:
ditto on the mutex

Reply:
for the batch producer the verifybatch is basically `return true`, hence you can get rid of this alltogether.